messaging/src/main/java/org/axonframework/messaging/GenericMessage.java

```diff
@@ -56,6 +57,23 @@
         }
     }
 
+    /**
+     * Extracts the message name from the given {@code payloadOrMessage}. If the given {@code payloadOrMessage} is an
+     * instance of {@link Message}, {@link Message#getPayloadType()} is used. Otherwise we expect the given {@link
+     * Object} to be the payload directly. The return value will be the result of {@link Class#getName()}, either from
+     * the {@code Message}'s {@code payloadType} or from the payload's class.
+     *
+     * @param payloadOrMessage the payload to extract the {@link Class#getName()} from, or an instance of {@link
+     *                         Message}
+     * @return the {@link Class#getName()} result from the payload, or from the {@link Message#getPayloadType()} if the
+     * given {@code payloadOrMessage} is an instance of {@link Message}
+     */
+    public static String messageName(Object payloadOrMessage) {
```

Comment:
I think this class also needs to take instance of `QueryMessage` into account, and use the `queryName` property to get the name.
Also suggest using `queryName` instead of `messageName` as method name and moving it to `GenericQueryMessage`.

Reply:
Good point. Moved this method to the `QueryMessage` interface, renamed it to `queryName` and added a third option: that the given `payloadOrMessage` can be of type `QueryMessage`, meaning `QueryMessage#getQueryName()` should be returned.